File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        result = list()

        for asteroid in asteroids:
            if (not result or
                result[-1] * asteroid > 0 or
                result[-1] < 0 and asteroid > 0):
                result.append(asteroid)
            else:
                flag = True
                while result and result[-1] * asteroid < 0:
                    if abs(result[-1]) < abs(asteroid):
                        result.pop()
                    elif abs(result[-1]) == abs(asteroid):
                        result.pop()
                        flag = False
                        break
                    else:
                        flag = False
                        break
                if flag:
                    result.append(asteroid)

        return result

File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> res = new Stack<>();

        for (int asteroid: asteroids) {
            if (res.isEmpty() ||
                res.peek() * asteroid > 0 ||
                res.peek() < 0 && asteroid > 0) {
                res.push(asteroid);
            } else {
                boolean flag = true;
                while (!res.isEmpty() && res.peek() * asteroid < 0) {
                    if (Math.abs(res.peek()) < Math.abs(asteroid)) {
                        res.pop();
                    } else if (Math.abs(res.peek()) == Math.abs(asteroid)) {
                        res.pop();
                        flag = false;
                        break;
                    } else {
                        flag = false;
                        break;
                    }
                }
                if (flag) {
                    res.push(asteroid);
                }
            }
        }

        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}